chanjo/store/api.py

```diff
@@ -0,0 +1,348 @@
+
+    Args:
+        query (Query): SQLAlchemy query object
+        group_id (Optional[str]): sample group identifier
+        sample_ids (Optional[List[str]]): sample ids
+
+    Returns:
+        Query: filtered query object
+    """
+    if group_id:
+        logger.debug('filter based on group')
+        return query.filter(Sample.group == group_id)
+    elif sample_ids:
+        logger.debug('filter based on list of samples')
+        return query.filter(Sample.sample_id.in_(sample_ids))
+    else:
+        return query
+
+
+class ChanjoAPI(Store):
+
+    """Interface to chanjo database for asking interesting questions.
+
+    Attributes:
+        weighten_average (BinaryExpression): weighter mean for metrics
+    """
+
+    def init_app(self, app, key_base='CHANJO_'):
+        """Configure API (Flask style) after lazy initialization.
+
+        Args:
+            app (Flask): Flask app instance
+            key_base (str): namespace to look for under ``app.config``
+
+        Returns:
+            ChanjoAPI: ``self``
+        """
+        uri = app.config["{}URI".format(key_base)]
+        self.connect(db_uri=uri)
+        return self
+
+    @property
+    def weighted_average(self):
+        """Build a dynamic field to calculate a weighted average.
+
+        Works for any of the exon metrics.
+
+        Returns:
+            BinaryExpression: dynamic field for a query
+        """
+        # length of the exon
+        weight = Exon.end - Exon.start
+        total_weight = func.sum(weight)
+        total_value = func.sum(ExonStatistic.value * weight)
+        weighted_mean = total_value / total_weight
+        return weighted_mean
+
+    def samples(self, group_id=None, sample_ids=None):
+        """Get samples from the database.
+
+        Args:
+            group_id (Optional[str]): sample group identifier
+            sample_ids (Optional[List[str]]): sample ids
+
+        Returns:
+            List[Sample]: samples objects from the database
+        """
+        query = self.query(Sample).order_by(Sample.sample_id)
+        query = filter_samples(query, group_id=group_id, sample_ids=sample_ids)
+        return query
+
+    def mean(self, *samples):
+        r"""Calculate mean values for all metrics on a per sample basis.
+
+        Args:
+            \*samples (Optional[List[str]]): filter by sample id
+
+        Returns:
+            dict: weighted averages grouped by sample
+        """
+        results = (self.query(Sample.sample_id, ExonStatistic.metric,
+                              self.weighted_average)
+                       .join(ExonStatistic.sample, ExonStatistic.exon)
+                       .group_by(Sample.sample_id, ExonStatistic.metric))
+        if samples:
+            results = results.filter(Sample.sample_id.in_(samples))
+
+        sample_groups = group_by_field(results, name='sample_id')
+        return sample_groups
+
+    def region(self, chromosome, start, end, sample_id=None, per=None):
```

Comment:
Could it be possible to use a region id here to? As an alternative to coordinates?